README.md
# java.io

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![codecov](https://codecov.io/gh/node-modules/java.io/branch/master/graph/badge.svg)](https://codecov.io/gh/node-modules/java.io)
[![David][david-image]][david-url]

[npm-image]: https://img.shields.io/npm/v/java.io.svg?style=flat
[npm-url]: https://npmjs.org/package/java.io
[travis-image]: https://img.shields.io/travis/node-modules/java.io.svg?style=flat
[travis-url]: https://travis-ci.org/node-modules/java.io
[david-image]: https://img.shields.io/david/node-modules/java.io.svg?style=flat
[david-url]: https://david-dm.org/node-modules/java.io

A node implement of "java.io.InputObjectStream.readObject()" and "java.io.OutputObjectStream.writeObject()".

## Protocol

- [Object Serialization Stream Protocol](http://docs.oracle.com/javase/6/docs/platform/serialization/spec/protocol.html) (A protocal is not all, still need to refer to the implementation.)
- [Object Serialization Stream Protocol Mind Node](https://www.dropbox.com/s/vfom3gz5t13h3w6/Object%20Serialization%20Stream%20Protocol.pdf)


## Install

```bash
$ npm install java.io --save
```

## Usage

### 1. readObject()

```js
var fs = require('fs');
var io = require('java.io');
var InputObjectStream = io.InputObjectStream;
var OutputObjectStream = io.OutputObjectStream;

// Read object and return whole info
var buf = fs.readFileSync('./test/fixtures/out/int/1024.bin');
var in = new InputObjectStream(buf, true);
var obj = in.readObject();

// Read object but return value only
var buf2 = fs.readFileSync('./test/fixtures/out/int/1024.bin');
var in2 = new InputObjectStream(buf);
var obj2 = in.readObject();
```

`obj` should be:

```js
{
  '$' : {
    value : 1024
  },
  '$class' : {
    name : 'java.lang.Integer',
    serialVersionUID : '1360826667806852920',
    flags : 2,
    fields : [{
      type : 'I',
      name : 'value'
    }],
    superClass : {
      name : 'java.lang.Number',
      serialVersionUID : '-8742448824652078965',
      flags : 2,
      fields : [],
      superClass : null
    }
  }
}
```

`obj2` should be:

```js
1024
```

if you only care about the first object from input stream, you could write the code briefly:

```js
var buf3 = fs.readFileSync('./test/fixtures/out/map/boolean.bin');
var obj3 = InputObjectStream.readObject(buf);
```

then `obj3` should be:

```js
{ 'true': true, 'false': false }
```

### 2. writeObject(obj)

```js
var outputObjectStream = new OutputObjectStream();

// 1. Passed in argument must contains the whole info
// 2. Every time calling the writeObject function
//    will return the buf had written in
var buf = outputObjectStream.writeObject(obj);
```

A brief style is also OK:

```js
OutputObjectStream.writeObject(obj);
```

### 3. OutputObjectStream.normalize(obj, type)

A convenient way to convert ordinary JavaScript object to object of standard format with whole info.

- params
  - obj: accept all primitive value or primitive array and map
  - type: string | boolean | int | short | long | char | byte | float | double |
- return: normalized object

```js
var outputObjectStream = new OutputObjectStream();
var normalizedObj = OutputObjectStream.normalize(true);
var buf = outputObjectStream.writeObject(normalizedObj);
```

```js
normalize(null)

normalize('string')

normalize(true)

normalize(1) // quals to normalize(1, 'int')

normalize(-123456, long)

normalize([ true, false, false, false ], 'boolean')

normalize( {'true': true, 'false': false}, 'boolean')
```

### 4. addObject()

If a class has writeObject/readObject methods, you need to implement the corresponding methods, and add them via addObject() before read or write the object.

```js
var io = require('java.io');
io.addObject({{className}}, {{class}});

```

Builtin classes:

- [java.util.ArrayList](./lib/objects/array_list.js)
- [java.util.LinkedList](./lib/objects/linked_list.js)
- [java.util.HashMap](./lib/objects/hash_map.js)
- [java.util.HashSet](./lib/objects/hash_set.js)
- [java.util.TreeMap](./lib/objects/tree_map.js)
- [java.util.TreeSet](./lib/objects/tree_set.js)
- [java.util.Date](./lib/objects/date.js)

## Data structure

```js
{
  // if a object has it's own readObject/writeObject method
  // save it's special value here
  '_$': ...,

  // value of object
  '$': ...,

  // class description
  '$class': {
    name: 'className',
    serialVersionUID: 'SVUID',
    flags: flags,
    fields:
     [ { type: 'F', name: 'primitiveProperty' },
       { type: 'L', name: 'objProperty', classname: 'Ljava/lang/String;' }],
    superClass: parentClassDescriptionOrNull
  }
}
```

[Some more examples](test/fixtures/in/)

## Incompatible between 1.x and 2.x

- decode java `[B` to `new Buffer([1, 2, 3])` not `[1, 2, 3]` [#10](https://github.com/node-modules/java.io/pull/10)

## License

[MIT](LICENSE)


History.md

2.4.0 / 2019-01-08
==================

**features**
  * [[`afcfb48`](http://github.com/node-modules/java.io/commit/afcfb4881ba18c72f86bd7dd453f523a371e3e35)] - feat: dump buf while decode error occurred (#27) (zōng yǔ <<gxcsoccer@users.noreply.github.com>>)

2.3.0 / 2016-11-29
==================

  * feat: support java.util.TreeMap (#21)

2.2.0 / 2016-11-28
==================

  * feat: support more Builtin class (#20)

2.1.2 / 2016-06-08
==================

  * chore: fix Object Serialization Stream Protocol Mind Node link
  * deps: upgrade all deps
  * chore: remove unuse comments
  * fix: write longer bytes block header

2.1.1 / 2015-10-28
==================

 * chore: replace const to var, to support old version node

2.1.0 / 2015-06-04
==================

 * feat: support read large obj bytes

2.0.1 / 2014-12-02
==================

 * fix(output): support write undefined property

2.0.0 / 2014-12-01
==================

 * feat(input): readObject byte -> Buffer, incompatible feature

1.2.4 / 2014-11-24
==================

 * use npm scripts instead of Makefile
 * fix get superClass bug. Closes #8

1.2.3 / 2014-11-16
==================

 * fix: try to detect class have readObject or not

1.2.2 / 2014-08-19
==================

 * Add java.util.LinkedList.
 * Add more helpers.

1.2.0 / 2014-08-06
==================

 * Tighten the check if an class that has read/write object method has been added in.
 * Update docs on addObject method.
 * Convert input.addObject() and output.addObject() to io.addObject()
 * Give a distinct error message when some class dose not implement its readObject() or writeObject()

1.1.1 / 2014-08-05
==================

 * Fix read field of array type error.
 * Check if field of array type is null first when get fields.

1.1.0 / 2014-08-04
==================

 * Fix jshint errors.
 * Update _writeSerialData()
 * Remove unused param: unshared.
 * Read/writeBlockHeader in class that has its own read/writeObject method.
 * Object overwrite read/writeObject() will call defaultRead/WriteObject() in its methods.
 * fix string
 * update readme code style

1.0.0 / 2014-07-31
==================

 * Impl Object Serialization Stream Protocol v2.0 (@fool2fish)

0.0.1 / 2014-04-04
==================

 * _readHandle() start
 * fix Object[] array
 * readObject as ArrayList<String>
 * readObject as String
 * support Object[] params
 * support Serializable String, Boolean, null, simple object
 * init


lib/normalize.js
'use strict';

var numCls = {
  name : 'java.lang.Number',
  serialVersionUID : '-8742448824652078965',
  flags : 2,
  fields : [],
  superClass : null
}

var primitiveClsDesc = {

  'boolean' : {
    name : 'java.lang.Boolean',
    serialVersionUID : '-3665804199014368530',
    flags : 2,
    fields : [{
      type : 'Z',
      name : 'value'
    }],
    superClass : null
  },

  'char' : {
    name : 'java.lang.Character',
    serialVersionUID : '3786198910865385080',
    flags : 2,
    fields : [{
      type : 'C',
      name : 'value'
    }],
    superClass : null
  },

  'byte' : {
    name : 'java.lang.Byte',
    serialVersionUID : '-7183698231559129828',
    flags : 2,
    fields : [{
      type : 'B',
      name : 'value'
    }],
    superClass : numCls
  },

  'int' : {
    name : 'java.lang.Integer',
    serialVersionUID : '1360826667806852920',
    flags : 2,
    fields : [{
      type : 'I',
      name : 'value'
    }],
    superClass : numCls
  },

  'short' : {
    name : 'java.lang.Short',
    serialVersionUID : '7515723908773894738',
    flags : 2,
    fields : [{
      type : 'S',
      name : 'value'
    }],
    superClass : numCls
  },

  'long' : {
    name : 'java.lang.Long',
    serialVersionUID : '4290774380558885855',
    flags : 2,
    fields : [{
      type : 'J',
      name : 'value'
    }],
    superClass : numCls
  },

  'float' : {
    name : 'java.lang.Float',
    serialVersionUID : '-2671257302660747028',
    flags : 2,
    fields : [{
      type : 'F',
      name : 'value'
    }],
    superClass : numCls
  },

  'double' : {
    name : 'java.lang.Double',
    serialVersionUID : '-9172774392245257468',
    flags : 2,
    fields : [{
      type : 'D',
      name : 'value'
    }],
    superClass : numCls
  }
}

var arrayClsDesc = {

  'boolean' : {
    name: '[Z',
    serialVersionUID: '6309297032502205922',
    flags: 2,
    fields: [],
    superClass: null
  },

  'char' : {
    name : '[C',
    serialVersionUID : '-5753798564021173076',
    flags : 2,
    fields : [],
    superClass : null
  },

  'byte' : {
    name: '[B',
    serialVersionUID: '-5984413125824719648',
    flags: 2,
    fields: [],
    superClass: null
  },

  'int' : {
    name: '[I',
    serialVersionUID: '5600894804908749477',
    flags: 2,
    fields: [],
    superClass: null
  },

  'short' : {
    name: '[S',
    serialVersionUID: '-1188055269542874886',
    flags: 2,
    fields: [],
    superClass: null
  },

  'long' : {
    name: '[J',
    serialVersionUID: '8655923659555304851',
    flags: 2,
    fields: [],
    superClass: null
  },

  'float' : {
    name: '[F',
    serialVersionUID: '836686056779680834',
    flags: 2,
    fields: [],
    superClass: null
  },

  'double' : {
    name: '[D',
    serialVersionUID: '4514449696888150558',
    flags: 2,
    fields: [],
    superClass: null
  },

  'string' : {
    name : '[Ljava.lang.String;',
    serialVersionUID : '-5921575005990323385',
    flags : 2,
    fields : [],
    superClass : null
  }
}

var mapCls = {
  name: 'java.util.HashMap',
  serialVersionUID: '362498820763181265',
  flags: 3,
  fields:
   [ { type: 'F', name: 'loadFactor' },
     { type: 'I', name: 'threshold' } ],
  superClass: null
}


/**
 * @param obj
 * @param [type]
 * @return        standard object with whole info
 */
function normalize(obj, type) {
  var ret = obj;
  if (obj === null || typeof obj === 'string') {
    ret = obj;
  } else if (typeof obj === 'boolean') {
    ret = {
      '$' : { value : obj },
      '$class' : primitiveClsDesc.boolean
    }
  } else if (typeof obj === 'number') {
    var isInt = Math.ceil(obj) === Math.floor(obj);
    if (!isInt && !type) {
      throw new Error('Cannot dicide obj type');
    }
    type = type || 'int';

    if (type in primitiveClsDesc) {
      ret = {
        '$' : { value : obj },
        '$class' : primitiveClsDesc[type]
      }

    } else {
      throw new Error('Illegal obj type');
    }
  } else if (Array.isArray(obj)) {
    if (!type) {
      throw new Error('Argument type is required when pass an array in');
    }

    if (type in arrayClsDesc) {
      ret = {
        '$' : obj,
        '$class' : arrayClsDesc[type]
      }
    } else {
      throw new Error('Cannot handle non-primitive array');
    }
  } else if (obj.toString() === '[object Object]') {
    if (!type) {
      throw new Error('Argument type is required when pass an object in');
    }

    if (type in primitiveClsDesc || type === 'string') {
      var values = {};
      for (var i in obj) {
        values[i] = normalize(obj[i], type);
      }
      ret = {
        '_$' : values,
        '$': { loadFactor: 0.75, threshold: getThreshold(obj)},
        '$class' : mapCls
      }
    } else {
      throw new Error('Cannot handle non-primitive object');
    }
  } else {
    throw new Error('Cannot handle obj');
  }

  return ret;
}


function getThreshold(obj) {
  var len = Object.keys(obj).length;
  var ret = Math.ceil(len / 8) * 6;  //*Math.ceil(amount / 8) * 8 * 0.75*/
  return ret < 12 ? 12 : ret
}

module.exports = normalize;
